File: src/main/java/com/everyonewaiter/application/auth/service/AuthService.java
package com.everyonewaiter.application.auth.service;

import com.everyonewaiter.application.auth.service.request.SendAuthCode;
import com.everyonewaiter.application.auth.service.request.VerifyAuthCode;
import com.everyonewaiter.domain.auth.entity.AuthAttempt;
import com.everyonewaiter.domain.auth.entity.AuthCode;
import com.everyonewaiter.domain.auth.entity.AuthPurpose;
import com.everyonewaiter.domain.auth.entity.AuthSuccess;
import com.everyonewaiter.domain.auth.event.AuthCodeSendEvent;
import com.everyonewaiter.domain.auth.event.AuthMailSendEvent;
import com.everyonewaiter.domain.auth.repository.AuthRepository;
import com.everyonewaiter.domain.auth.service.AuthValidator;
import com.everyonewaiter.global.security.JwtPayload;
import com.everyonewaiter.global.security.JwtProvider;
import java.time.Duration;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

  private final JwtProvider jwtProvider;
  private final AuthValidator authValidator;
  private final AuthRepository authRepository;
  private final ApplicationEventPublisher applicationEventPublisher;

  public void checkExistsAuthSuccess(String phoneNumber, AuthPurpose purpose) {
    AuthSuccess authSuccess = new AuthSuccess(phoneNumber, purpose);
    authValidator.checkExistsAuthSuccess(authSuccess);
  }

  public void sendAuthCode(SendAuthCode request) {
    AuthAttempt authAttempt = new AuthAttempt(request.phoneNumber(), request.purpose());
    authValidator.validateAuthAttempt(authAttempt);

    AuthCode authCode = new AuthCode(request.phoneNumber());
    authRepository.save(authCode);
    authRepository.increment(authAttempt);
    authRepository.delete(new AuthSuccess(request.phoneNumber(), request.purpose()));

    AuthCodeSendEvent event = new AuthCodeSendEvent(request.phoneNumber(), authCode.code());
    applicationEventPublisher.publishEvent(event);
  }

  public void verifyAuthCode(VerifyAuthCode request) {
    AuthSuccess authSuccess = new AuthSuccess(request.phoneNumber(), request.purpose());
    authValidator.checkNotExistsAuthSuccess(authSuccess);

    AuthCode authCode = new AuthCode(request.phoneNumber(), request.code());
    int code = authRepository.find(authCode);
    authCode.verify(code);

    authRepository.save(authSuccess);
    authRepository.delete(authCode);
  }

  public String generateToken(JwtPayload payload, Duration expiration) {
    return jwtProvider.generate(payload, expiration);
  }

  @Transactional(readOnly = true)
  public void sendAuthMail(String email) {
    applicationEventPublisher.publishEvent(new AuthMailSendEvent(email));
  }

}

File: src/main/kotlin/com/everyonewaiter/application/auth/service/AuthService.kt
package com.everyonewaiter.application.auth.service

import com.everyonewaiter.application.auth.dto.VerifyAuthCode
import com.everyonewaiter.common.jwt.JwtPayload
import com.everyonewaiter.common.jwt.JwtProvider
import com.everyonewaiter.domain.auth.entity.AuthAttempt
import com.everyonewaiter.domain.auth.entity.AuthCode
import com.everyonewaiter.domain.auth.entity.AuthPurpose
import com.everyonewaiter.domain.auth.entity.AuthSuccess
import com.everyonewaiter.domain.auth.event.AuthCodeCreateEvent
import com.everyonewaiter.domain.auth.event.AuthMailSendEvent
import com.everyonewaiter.domain.auth.repository.AuthAttemptRepository
import com.everyonewaiter.domain.auth.repository.AuthCodeRepository
import com.everyonewaiter.domain.auth.repository.AuthSuccessRepository
import com.everyonewaiter.global.exception.ErrorCode
import com.everyonewaiter.global.extension.checkOrThrow
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Service

@Service
class AuthService(
    private val jwtProvider: JwtProvider,
    private val authAttemptRepository: AuthAttemptRepository,
    private val authCodeRepository: AuthCodeRepository,
    private val authSuccessRepository: AuthSuccessRepository,
    private val applicationEventPublisher: ApplicationEventPublisher,
) {
    fun checkAuthSuccessExists(phoneNumber: String) {
        checkOrThrow(existsSuccess(phoneNumber), ErrorCode.EXPIRED_VERIFICATION_PHONE_NUMBER)
    }

    fun checkAuthSuccessNotExists(phoneNumber: String) {
        checkOrThrow(existsSuccess(phoneNumber).not(), ErrorCode.ALREADY_VERIFIED_PHONE_NUMBER)
    }

    private fun existsSuccess(phoneNumber: String) = authSuccessRepository.exists(AuthSuccess(phoneNumber))

    fun checkAuthAttemptExceed(
        phoneNumber: String,
        purpose: AuthPurpose,
    ) {
        val authAttempt = AuthAttempt(phoneNumber, purpose)
        val isExceed = authAttemptRepository.find(authAttempt) < purpose.maxAttempt
        checkOrThrow(isExceed, ErrorCode.EXCEED_MAXIMUM_VERIFICATION_PHONE_NUMBER)
    }

    fun generateAccessTokenBySignIn(
        accountId: Long,
        email: String,
    ): String = jwtProvider.generate(JwtPayload(accountId.toString(), email))

    fun generateCode(
        phoneNumber: String,
        purpose: AuthPurpose,
    ) {
        val authCode = AuthCode(phoneNumber)
        authCodeRepository.save(authCode)
        authAttemptRepository.increment(AuthAttempt(phoneNumber, purpose))
        applicationEventPublisher.publishEvent(AuthCodeCreateEvent(phoneNumber, authCode.code))
    }

    fun verifyCode(request: VerifyAuthCode.Request) {
        val code = authCodeRepository.find(AuthCode(request.phoneNumber))
        checkOrThrow(code != null, ErrorCode.EXPIRED_VERIFICATION_CODE)
        checkOrThrow(code == request.code, ErrorCode.UNMATCHED_VERIFICATION_CODE)
        authSuccessRepository.save(AuthSuccess(request.phoneNumber))
    }

    fun sendAuthMail(email: String) {
        applicationEventPublisher.publishEvent(AuthMailSendEvent(email))
    }
}
